Cleanup/gf test interface

cpviolator opened the pull request:
This PR adds CL params for the gauge fixing routines so that one can control the parameters. The original gauge alg test is now a `ctest` with the Google testing suite, and the file `gauge_alg_test.cpp` performs only a single gauge fix test specified by the user, with the option to save the gauge field.

cpviolator committed:
Add gf test interface to allow fine grained control over the GF testing

cpviolator committed:
Add gf test interface to allow fine grained control over the GF testing

cpviolator committed:
Move the gauge al test to a ctest, make a new interface to the gauge fixing that allows for fine grained control and gauge IO

cpviolator committed:
Move the gauge_alg_test to a ctest, make a new interface to the gauge fixing that allows for fine grained control and gauge IO.

cpviolator committed:
Move the gauge_alg_test to a ctest, make a new interface to the gauge fixing that allows for fine grained control and gauge IO.

cpviolator committed:
Move the gauge al test to a ctest, make a new interface to the gauge fixing that allows for fine grained control and gauge IO

cpviolator committed:
Merge branch 'cleanup/gf_test_interface' of https://github.com/lattice/quda into cleanup/gf_test_interface

cpviolator committed:
clang tidy

cpviolator committed:
clang tidy

cpviolator committed:
Allow for single case testing in gauge_alg_ctest, minor clean up of gauge fixing stdout and comments

cpviolator committed:
Allow for single case testing in gauge_alg_ctest, minor clean up of gauge fixing stdout and comments

cpviolator committed:
Clang tidy

cpviolator committed:
Clang tidy

maddyscientist reviewed (commented):
This is a solid PR, but given we're working on merging GK into develop soon, would you mind rebasing this against GK?  Apologies for the churn.

cpviolator commented:
Will do, no problem at all.

On Thu, Jul 29, 2021 at 10:57 AM maddyscientist ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
>
> This is a solid PR, but given we're working on merging GK into develop
> soon, would you mind rebasing this against GK? Apologies for the churn.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/lattice/quda/pull/1146#pullrequestreview-718381880>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAR7TV3PR6SJ35USBOBLZ5TT2GI67ANCNFSM435ZH3JQ>
> .
>

cpviolator committed:
Rebase to GK

cpviolator commented:
Rebased this against GK, builds and runs fine. I moved the `gauge_alg_test.cpp` to `gauge_alg_ctest.cpp` as it is incorporated into the google testing suite.

cpviolator committed:
Add gauge param argument

cpviolator committed:
GK merge

maddyscientist commented:
This PR needs develop merged in, and then we can review this.

cpviolator committed:
mid merge

cpviolator committed:
Even with develop

cpviolator commented:
I see some regression in the FFT gauge fixing. It is present in the current develop branch too for both tuned and untuned runs. The reported data in the FFT gauge fixing run does not make sense in that the action is incorrect and there is no convergence. If allowed to complete to the max iterations, the plaquette value is the same as it was prior to gauge fixing, so the routine is not leaving the gauge orbit. I've pasted a command and the full output is attached. If I can't find a solution I'll open an issue for the develop branch.

```
./gauge_alg_test --dim 8 8 8 8 --gf-tol 1e-10 --gf-maxiter 1000
```

[github_gf_output.txt](https://github.com/lattice/quda/files/7324604/github_gf_output.txt)

cpviolator committed:
Make gf_fft_autotune default to true

cpviolator commented:
The issue is that the default value passed for `--gf-fft-autotune` should be `true` else the default `alpha` value for the lattice may be poor. Fixed with https://github.com/lattice/quda/commit/bb9c0de72f285c002fb140f9a2bae100d80f3cd6

cpviolator committed:
Remove alpha autotuning in FFT gauge fixing, add comments where appropriate. Fix gauge dir error in gauge_alg_test, adjust step output in gauge fixing

maddyscientist commented on tests/CMakeLists.txt:
```diff
@@ -812,8 +812,8 @@ foreach(prec IN LISTS TEST_PRECS)
-             COMMAND ${QUDA_CTEST_LAUNCH} $<TARGET_FILE:gauge_alg_test> ${MPIEXEC_POSTFLAGS}
-                     --dim 4 6 8 10 --prec ${prec}
+             COMMAND ${QUDA_CTEST_LAUNCH} $<TARGET_FILE:gauge_alg ctest> ${MPIEXEC_POSTFLAGS}
+                     --dim 2 4 6 8 --prec ${prec}
```
This size needs to be reverted I think: I intentionally made the ctest problem size 4x6x8x10.  If I recall correctly, one cannot run multi-GPU with a local dimension of size 2: the issue is that there is no interior, since everything is halo, and you get failures as a result.

maddyscientist replied:
Also, this looks like a type: `gauge_alg ctest` and not `gauge_alg_ctest`

cpviolator replied:
https://github.com/lattice/quda/commit/040be93004b847b2092280c1b79016904943a511

maddyscientist commented on tests/gauge_alg_ctest.cpp:
```diff
@@ -293,7 +328,7 @@ class GaugeAlgTest : public ::testing::Test
 
     void *cpu_gauge[4];
-    for (int dir = 0; dir < 4; dir++) { cpu_gauge[dir] = safe_malloc(V * gauge_site_size * gauge_param.cpu_prec); }
+    for (int dir = 0; dir < 4; dir++) { cpu_gauge[dir] = malloc(V * gauge_site_size * gauge_param.cpu_prec); }
```
This should be changed back to `safe_malloc`: we should be tracking allocations in the tests

cpviolator replied:
Done

maddyscientist commented on tests/gauge_alg_ctest.cpp:
```diff
@@ -69,37 +60,75 @@ void display_test_info()
-
-  Timer<false> a0, a1;
+  
+  Timer<false> a0,a1;
```
Perhaps we should switch this use of `Timer` to `host_timer_t`.  See dslash_test for example.

cpviolator replied:
I used `device_timer_t` as there is no host cross check code.

maddyscientist commented on lib/gauge_fix_fft.cu:
```diff
@@ -185,14 +185,26 @@ namespace quda {
+    // autotuning ensures optimal behaviour.
+    // Users who wish to change this behaviour may remove the follwing line
+    // of code and recompile to regain control of alpha autotuning.
+    autotune = 1;
```
I think this hard coding of the autotune parameter makes me uneasy.

What we could do is replace the interface for `computeGaugeFixingFFTQuda` and `computeGaugeFixingOVRQuda`:  instead of passing the arguments all separately to the interface function, we should make a interface param, e.g., `QudaGaugeFixParam` and place all the gauge fixing parameters in such a struct.  This would allow us to have default values if we create partner helper functions `initQudaGaugeFixParam` (as well as `printQudaGaugeFixParam` and `checkQudaGaugeFixParam`) in check_param.h.

What do you think @cpviolator ?

cpviolator replied:
What is the best approach with regards to the interface function signatures for gauge fixing? If we go ahead and change them, we may as well consolidate the two functions into a single call to `computeQudaGaugeFixing`, add a `QUDA_GAUGEFIX_TYPE` and direct the workflow within it. We could preserve the two existing function signatures too as wrappers to this consolidated function for back compatibility, or force downstream modifications.

cpviolator replied:
As far as I can tell, the MILC interface would be largely unaffected as we simply change the inner plumbings of
```C++
void qudaGaugeFixingOVR(...)
```
and
```C++
void qudaGaugeFixingFFT(...)
```

maddyscientist replied:
Having a single gauge fixing interface function sounds like a great idea.  Agree regarding MILC itself not needing to be changed, since it goes through milc_interface.cpp, so we can keep these changes internal to QUDA.

Since this breaking change will align with QUDA 2.0, e.g., a major version number change, I think there's no need to maintain the prior legacy interface as a fallback.  So I think we can just remove the prior interface.

cpviolator replied:
https://github.com/lattice/quda/commit/aed7763076cef871a73157f6c358e2a053ec6c83

maddyscientist reviewed (changes requested):
Thanks for this PR @cpviolator! I've given it a visual review, and left comments inline.  This would be a nice update to the gauge fixing.

cpviolator commented:
Thanks Kate, I'll fix all of these and clang tidy. I agree RE the gauge fixing param, it's a neat way to address bloating function signatures and control defaults.

cpviolator committed:
Merge remote-tracking branch 'origin/develop' into cleanup/gf_test_interface

cpviolator committed:
Revert MPI test params

cpviolator committed:
Use device_timer instead of Timer

cpviolator committed:
Remove extra semi-colon

cpviolator committed:
Consolidate the gauge fixing interafce, use new parameter struct QudaGaugeFixParam to define both parameters and algorithm, format variable names to be consistent, use QudaBoolean for booleans rather than integers

cpviolator committed:
Clang format tidy using clang-format-quda-11 develop cleanup/gf_test_interface on NVSOCAL2

cpviolator commented:
All comments addressed. I use `QudaBoolean` for the bools in the `fix.cu` files to remain consistent with the interface, but I can convert them to `bool` prior to the device function call is that is not the desired method.

cpviolator commented:
BTW, this will need checking with the MILC interface for completeness. Google testing:
```
mpirun -np 2 ./gauge_alg_ctest --dim 8 8 8 8 --test 0 --gridsize 1 1 1 2
```
and spot testing
```
mpirun -np 2 ./gauge_alg_ctest --dim 8 8 8 8 --save-gauge test --test [1/2] --gridsize 1 1 1 2
```

cpviolator committed:
Revert to using host_timer_t in gauge_alg_ctest.cpp

cpviolator committed:
Use up-to-date local version of develop for clang tidy

cpviolator committed:
More formatting

cpviolator committed:
Clang from untouched files

cpviolator committed:
Clang with clean directory
